common/playlist.c

```diff
@@ -178,6 +179,27 @@ void playlist_shuffle(struct playlist *pl)
     talloc_free(arr);
     pl->current = save_current;
     pl->current_was_replaced = save_replaced;
+    pl->shuffled = true;
+}
+
+void playlist_unshuffle(struct playlist *pl) {
+    struct playlist_entry *save_current = pl->current;
+    bool save_replaced = pl->current_was_replaced;
+    int count = playlist_count(pl);
+    struct playlist_entry **arr = talloc_array(NULL, struct playlist_entry *,
+                                               count);
+    for (int n = 0; n < count; n++) {
+        arr[pl->first->pos] = pl->first;
```

Comment:
> I wasn't sure how safe this function should be, because my thought was, this function wouldn't be called unless pl->shuffled is true and thus pl->pos should be defined.

New entries could be added or removed or reordered after shuffle any time. It also could be shuffled again at any point. So your code has to handle that somehow, without exploding or leaking.